torch/csrc/jit/codegen/cuda/lower_warp_reduce.cpp

```diff
@@ -0,0 +1,546 @@
+          std::make_unique<std::vector<kir::Allocate*>>());
+    }
+    for (auto expr : for_loop->body().exprs()) {
+      handle(expr);
+    }
+    if (open_nest_level) {
+      running_kir_tv_to_allocate_map_.pop_back();
+    }
+  }
+
+  void handle(kir::IfThenElse* ite) {
+    running_visible_allocation_stack_.emplace_back(
+        std::make_unique<std::vector<kir::Allocate*>>());
+    for (auto expr : ite->thenBody().exprs()) {
+      handle(expr);
+    }
+    running_visible_allocation_stack_.pop_back();
+    running_visible_allocation_stack_.emplace_back(
+        std::make_unique<std::vector<kir::Allocate*>>());
+    for (auto expr : ite->elseBody().exprs()) {
+      handle(expr);
+    }
+    running_visible_allocation_stack_.pop_back();
+  }
+
+  //! Place this allocate on the list of currently visible allocates,
+  //!  organized by loop nest level.
+  void handle(kir::Allocate* allocate) {
+    if (allocate->memoryType() != MemoryType::Local) {
+      return;
+    }
+    if (auto kir_tv = dynamic_cast<kir::TensorView*>(allocate->buffer())) {
+      auto fuser_tv = kir_tv->fuserTv();
+      if (fuser_tv->definition()) {
+        if (fuser_tv->definition()->isA<ReductionOp>() ||
+            fuser_tv->definition()->isA<BroadcastOp>()) {
+          running_visible_allocation_stack_.back()->push_back(allocate);
+        }
+      }
+    }
+  }
+
+  //! Checks if the given tv has been replaced by broadcast fusion.
+  //!  returns the replaced TensorIndex if so.
+  c10::optional<kir::TensorIndex*> findMaybeReplacedTensorIndex(
+      kir::TensorIndex* tensor_index) {
+    auto kir_tv = tensor_index->view();
+    auto tensor_index_it = running_tv_replacement_map_.find(kir_tv);
+    if (tensor_index_it != running_tv_replacement_map_.end()) {
+      return tensor_index_it->second;
+    }
+    return c10::nullopt;
+  }
+
+  //! Iteratve backwards on the currently visible loop scopes
+  //!  and find the first allocate corresponding to the
+  //!  given tv.
+  kir::Allocate* getActiveAllocateFor(kir::TensorView* tv) {
+    for (auto frame_it = running_visible_allocation_stack_.rbegin();
+         frame_it != running_visible_allocation_stack_.rend();
+         frame_it++) {
+      for (auto allocate_it = (*frame_it)->rbegin();
+           allocate_it != (*frame_it)->rend();
+           allocate_it++) {
+        auto candidate_allocate = *allocate_it;
+        if (candidate_allocate->buffer() == tv) {
+          return candidate_allocate;
+        }
+      }
+    }
+    TORCH_INTERNAL_ASSERT(
+        false, "lower_warp_reduce: cannot find allocate for this op");
+    return nullptr;
+  }
+
+  Expr* getFuserTVExpr(kir::Expr* expr) {
+    auto out = expr->outputs()[0];
+    auto out_ti = dynamic_cast<kir::TensorIndex*>(out);
+    if (!out_ti) {
+      return nullptr;
+    }
+    return out_ti->view()->fuserTv()->definition();
+  }
+
+  bool isOpInputRegisterTV(kir::Expr* expr) {
+    auto fuser_expr = getFuserTVExpr(expr);
```

Comment:
Do you need to look at Fuser expressions here? Is KIR not sufficient?

Reply:
Yes KIR is sufficient. Updated. Thanks!